coderd/parameters.go

```diff
@@ -59,72 +62,49 @@ func (api *API) templateVersionDynamicParameters(rw http.ResponseWriter, r *http
-	fileID, err := api.Database.GetFileIDByTemplateVersionID(fileCtx, templateVersion.ID)
-	if err != nil {
-		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
-			Message: "Internal error finding template version Terraform.",
-			Detail:  err.Error(),
-		})
-		return
-	}
-
-	templateFS, err := api.FileCache.Acquire(fileCtx, fileID)
-	if err != nil {
-		httpapi.Write(ctx, rw, http.StatusNotFound, codersdk.Response{
-			Message: "Internal error fetching template version Terraform.",
-			Detail:  err.Error(),
-		})
-		return
-	}
-	defer api.FileCache.Release(fileID)
-
-	// Having the Terraform plan available for the evaluation engine is helpful
-	// for populating values from data blocks, but isn't strictly required. If
-	// we don't have a cached plan available, we just use an empty one instead.
-	plan := json.RawMessage("{}")
 	tf, err := api.Database.GetTemplateVersionTerraformValues(ctx, templateVersion.ID)
-	if err == nil {
-		plan = tf.CachedPlan
-
-		if tf.CachedModuleFiles.Valid {
-			moduleFilesFS, err := api.FileCache.Acquire(fileCtx, tf.CachedModuleFiles.UUID)
-			if err != nil {
-				httpapi.Write(ctx, rw, http.StatusNotFound, codersdk.Response{
-					Message: "Internal error fetching Terraform modules.",
-					Detail:  err.Error(),
-				})
-				return
-			}
-			defer api.FileCache.Release(tf.CachedModuleFiles.UUID)
-			templateFS = files.NewOverlayFS(templateFS, []files.Overlay{{Path: ".terraform/modules", FS: moduleFilesFS}})
-		}
-	} else if !xerrors.Is(err, sql.ErrNoRows) {
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
 		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
 			Message: "Failed to retrieve Terraform values for template version",
 			Detail:  err.Error(),
 		})
 		return
 	}
 
-	// If the err is sql.ErrNoRows, an empty terraform values struct is correct.
+	// staticDiagnostics is a set of diagnostics to be applied to all rendered results.
 	staticDiagnostics := parameterProvisionerVersionDiagnostic(tf)
 
-	owner, err := api.getWorkspaceOwnerData(ctx, user, templateVersion.OrganizationID)
-	if err != nil {
-		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
-			Message: "Internal error fetching workspace owner.",
-			Detail:  err.Error(),
-		})
-		return
-	}
-
-	input := preview.Input{
-		PlanJSON:        plan,
-		ParameterValues: map[string]string{},
-		Owner:           owner,
+	// render is the function that given a set of input values, will return the
+	// parameter state. There is 2 rendering functions.
+	//
+	// prepareStaticPreview uses the static set of parameters saved from the template
+	// import. These parameters are returned on every request, and have no dynamic
+	// functionality. This exists for backwards compatibility with older template versions
+	// which have not uploaded their plan & module files.
+	//
+	// prepareDynamicPreview uses the dynamic preview engine.
+	var render previewFunction
```

Comment:
```suggestion
	var preview previewFunction
```

Reply:
or even better, why bother with all this indirection? just figure out which of the functions we should call and call it directly. this just makes it too hard to follow.